packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,436 @@
+  const FadeInImage({
+    Key key,
+    @required this.placeholder,
+    @required this.image,
+    this.fadeOutDuration: const Duration(milliseconds: 300),
+    this.fadeOutCurve: Curves.easeOut,
+    this.fadeInDuration: const Duration(milliseconds: 700),
+    this.fadeInCurve: Curves.easeIn,
+    this.width,
+    this.height,
+    this.fit,
+    this.alignment,
+    this.repeat: ImageRepeat.noRepeat,
+    this.centerSlice,
+  }) : assert(placeholder != null),
+       assert(image != null),
+       assert(fadeOutDuration != null),
+       assert(fadeOutCurve != null),
+       assert(fadeInDuration != null),
+       assert(fadeInCurve != null),
+       super(key: key);
+
+  /// Creates a widget that uses a placeholder image stored in-memory while
+  /// loading the final image from the network.
+  ///
+  /// [placeholder] contains the bytes of the in-memory image.
+  ///
+  /// [image] is the URL of the final image.
+  ///
+  /// The [placeholder], [image], [fadeOutDuration], [fadeOutCurve],
+  /// [fadeInDuration] and [fadeInCurve] arguments must not be null.
+  ///
+  /// See also:
+  ///
+  ///  * [new Image.memory], which has more details about loading images from
+  ///    memory.
+  ///  * [new Image.network], which has more details about loading images from
+  ///    the network.
+  FadeInImage.memoryNetwork({
+    Key key,
+    @required Uint8List placeholder,
+    @required String image,
+    double scale: 1.0,
+    this.fadeOutDuration: const Duration(milliseconds: 300),
+    this.fadeOutCurve: Curves.easeOut,
+    this.fadeInDuration: const Duration(milliseconds: 700),
+    this.fadeInCurve: Curves.easeIn,
+    this.width,
+    this.height,
+    this.fit,
+    this.alignment,
+    this.repeat: ImageRepeat.noRepeat,
+    this.centerSlice,
+  }) : assert(placeholder != null),
+      assert(image != null),
+      placeholder = new MemoryImage(placeholder, scale: scale),
+      image = new NetworkImage(image, scale: scale),
+      assert(fadeOutDuration != null),
+      assert(fadeOutCurve != null),
+      assert(fadeInDuration != null),
+      assert(fadeInCurve != null),
+      super(key: key);
+
+  /// Creates a widget that uses a placeholder image stored in an asset bundle
+  /// while loading the final image from the network.
+  ///
+  /// [placeholder] is the key of the image in the asset bundle.
+  ///
+  /// [image] is the URL of the final image.
+  ///
+  /// The [placeholder], [image], [fadeOutDuration], [fadeOutCurve],
+  /// [fadeInDuration] and [fadeInCurve] arguments must not be null.
+  ///
+  /// See also:
+  ///
+  ///  * [new Image.asset], which has more details about loading images from
+  ///    asset bundles.
+  ///  * [new Image.network], which has more details about loading images from
+  ///    the network.
+  FadeInImage.assetNetwork({
+    Key key,
+    @required String placeholder,
+    @required String image,
+    AssetBundle bundle,
+    double scale: 1.0,
```

Comment:
you should document the semantics of "scale" here
(and probably in the previous constructor too, but definitely here)